npe in schemacompat

```
E0819 13:26:56.469509       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 9771 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2d15dc0, 0x5383420})
        /go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220803191238-b6a732dbd013/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc02ef32700})
        /go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220803191238-b6a732dbd013/pkg/util/runtime/runtime.go:49 +0x75
panic({0x2d15dc0, 0x5383420})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/schemacompat.lcdForObject(0x1a, 0xc030c9bef0, 0xc030c9e000, 0xc030c9e0b0, 0x1)
        /workspace/pkg/schemacompat/schemacompat.go:364 +0x22d
github.com/kcp-dev/kcp/pkg/schemacompat.lcdForStructural(0xc030c90ba0, 0xc030c9bef0, 0xc030c9b760, 0x0, 0x70)
        /workspace/pkg/schemacompat/schemacompat.go:172 +0x125
github.com/kcp-dev/kcp/pkg/schemacompat.lcdForObject(0x413e85, 0xc030c9b6b0, 0xc030c9b760, 0xc030c9b810, 0x1)
        /workspace/pkg/schemacompat/schemacompat.go:357 +0xd25
github.com/kcp-dev/kcp/pkg/schemacompat.lcdForStructural(0xc030c90600, 0xc030c9b6b0, 0xc030c7e420, 0x0, 0xa0)
        /workspace/pkg/schemacompat/schemacompat.go:172 +0x125
github.com/kcp-dev/kcp/pkg/schemacompat.lcdForObject(0x0, 0xc030c7f130, 0xc030c7e420, 0xc001845290, 0x1)
        /workspace/pkg/schemacompat/schemacompat.go:357 +0xd25
github.com/kcp-dev/kcp/pkg/schemacompat.lcdForStructural(0xc030c7f130, 0xc030c7f130, 0x0, 0x0, 0x0)
        /workspace/pkg/schemacompat/schemacompat.go:172 +0x125
github.com/kcp-dev/kcp/pkg/schemacompat.EnsureStructuralSchemaCompatibility(0xc0381c15c0, 0xb, 0x0, 0x0)
        /workspace/pkg/schemacompat/schemacompat.go:69 +0x146
github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource.(*Controller).ensureAPIResourceCompatibility(0xc0062fb0e0, {0x375c778, 0xc030c903c0}, {{0xc0337e4b70, 0x40d5b4}}, {{0xc0337d8d20, 0x14}, {0xc0337a9e30, 0x8}, {0xc0337a9e60, ...}}, ...)
        /workspace/pkg/reconciler/apis/apiresource/negotiation.go:512 +0x1072
github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource.(*Controller).process(0xc0062fb0e0, {0x375c6d0, 0xc000151280}, {{0x3165c2b, 0x7}, {0x317bd1d, 0x11}, {0xc03637f320, 0x5e}, {{0xc0337d8d20, ...}, ...}, ...})
        /workspace/pkg/reconciler/apis/apiresource/negotiation.go:100 +0xd46
github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource.(*Controller).processNextWorkItem(0xc0062fb0e0, {0x375c6d0, 0xc000151280})
        /workspace/pkg/reconciler/apis/apiresource/controller.go:323 +0x1eb
github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource.(*Controller).startWorker(...)
        /workspace/pkg/reconciler/apis/apiresource/controller.go:307
github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource.(*Controller).Start.func1()
        /workspace/pkg/reconciler/apis/apiresource/controller.go:300 +0x45
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7f574fbd3b98)
        /go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220803191238-b6a732dbd013/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x746e6564693a6622, {0x36ffb80, 0xc0254135c0}, 0x1, 0xc0069abec0)
        /go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220803191238-b6a732dbd013/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc01800ffb0, 0x3b9aca00, 0x0, 0x0, 0x3a22656d6974222c)
        /go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220803191238-b6a732dbd013/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x938e6a, 0x726556697061227b, 0x61223a226e6f6973)
        /go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220803191238-b6a732dbd013/pkg/util/wait/wait.go:90 +0x25
created by github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource.(*Controller).Start
        /workspace/pkg/reconciler/apis/apiresource/controller.go:300 +0x1a5
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
```

https://github.com/kcp-dev/kcp/blob/65ef698085ce5c5f848fab4e174d531d4cdec6c4/pkg/schemacompat/schemacompat.go#L364

`new.AdditionalProperties` can be nil - need to guard against that

--- a/pkg/schemacompat/schemacompat.go
+++ b/pkg/schemacompat/schemacompat.go
@@ -24,7 +24,7 @@ import (
 
 	"go.uber.org/multierr"
 
-	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
 	"k8s.io/apimachinery/pkg/util/sets"
@@ -361,14 +361,14 @@ func lcdForObject(fldPath *field.Path, existing, new *schema.Structural, lcd *sc
 				delete(lcd.Properties, removedProperty)
 			}
 
-		} else if new.AdditionalProperties.Structural != nil {
+		} else if new.AdditionalProperties != nil && new.AdditionalProperties.Structural != nil {
 			for _, key := range sets.StringKeySet(existing.Properties).List() {
 				existingPropertySchema := existing.Properties[key]
 				lcdPropertySchema := lcd.Properties[key]
 				multierr.AppendInto(&err, lcdForStructural(fldPath.Child("properties").Key(key), &existingPropertySchema, new.AdditionalProperties.Structural, &lcdPropertySchema, narrowExisting))
 				lcd.Properties[key] = lcdPropertySchema
 			}
-		} else if new.AdditionalProperties.Bool {
+		} else if new.AdditionalProperties != nil && new.AdditionalProperties.Bool {
 			// that allows named properties only.
 			// => Keep the existing schemas as the lcd.
 		} else {

--- a/pkg/schemacompat/schemacompat_test.go
+++ b/pkg/schemacompat/schemacompat_test.go
@@ -20,6 +20,7 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+	"go.uber.org/multierr"
 
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	"k8s.io/apimachinery/pkg/util/validation/field"
@@ -440,12 +441,52 @@ func TestCompatibility(t *testing.T) {
 				},
 			},
 		},
+	}, {
+		desc: "existing has properties, new has neither properties or additionalProperties",
+		existing: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]apiextensionsv1.JSONSchemaProps{
+				"existing": {Type: "boolean"},
+			},
+		},
+		new: &apiextensionsv1.JSONSchemaProps{
+			Type: "array",
+			Items: &apiextensionsv1.JSONSchemaPropsOrArray{
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"existing": {Type: "integer"},
+					},
+				},
+			},
+		},
+		wantErr: multierr.Append(
+			field.Invalid(
+				field.NewPath("schema", "openAPISchema").Child("type"),
+				"array",
+				`The type changed (was "object", now "array")`,
+			),
+			field.Invalid(
+				field.NewPath("schema", "openAPISchema").Child("properties"),
+				[]string{"existing"},
+				"properties value has been completely cleared in an incompatible way",
+			),
+		),
 	}} {
 		t.Run(c.desc, func(t *testing.T) {
 			gotLCD, err := EnsureStructuralSchemaCompatibility(field.NewPath("schema", "openAPISchema"), c.existing, c.new, c.narrowExisting)
-			if d := cmp.Diff(c.wantErr, err); d != "" {
-				t.Errorf("Error Diff(-want,+got): %s", d)
+			if c.wantErr != nil {
+				if err == nil {
+					t.Fatalf("expected err %v but got nil", c.wantErr)
+				}
+
+				if d := cmp.Diff(c.wantErr.Error(), err.Error()); d != "" {
+					t.Errorf("Error Diff(-want,+got): %s", d)
+				}
+			} else if err != nil {
+				t.Fatalf("unexpected err %v", err)
 			}
+
 			if d := cmp.Diff(c.wantLCD, gotLCD); d != "" {
 				t.Errorf("LCD Diff(-want,+got): %s", d)
 			}

